{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd \n",
    "import numpy as np\n",
    "import df_helpers as helper\n",
    "from sanbomics.tools import id_map\n",
    "\n",
    "\n",
    "import os\n",
    "dirs = helper.dirs()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Construct gene info tables + significant DEGs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "ename": "FileNotFoundError",
     "evalue": "[Errno 2] No such file or directory: '1. Batch Counts/A_gene_counts.xlsx'",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mFileNotFoundError\u001b[0m                         Traceback (most recent call last)",
      "Cell \u001b[0;32mIn[3], line 4\u001b[0m\n\u001b[1;32m      1\u001b[0m DCTRL0_upsigs_df \u001b[38;5;241m=\u001b[39m helper\u001b[38;5;241m.\u001b[39mread_sigs(\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mUp\u001b[39m\u001b[38;5;124m'\u001b[39m,\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mDCTRL0\u001b[39m\u001b[38;5;124m'\u001b[39m)\n\u001b[1;32m      2\u001b[0m DCTRL0_downsigs_df \u001b[38;5;241m=\u001b[39m helper\u001b[38;5;241m.\u001b[39mread_sigs(\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mDown\u001b[39m\u001b[38;5;124m'\u001b[39m,\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mDCTRL0\u001b[39m\u001b[38;5;124m'\u001b[39m)\n\u001b[0;32m----> 4\u001b[0m batchA_df \u001b[38;5;241m=\u001b[39m \u001b[43mpd\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mread_excel\u001b[49m\u001b[43m(\u001b[49m\u001b[38;5;124;43mf\u001b[39;49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[38;5;132;43;01m{\u001b[39;49;00m\u001b[43mdirs\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mbatchcounts\u001b[49m\u001b[38;5;132;43;01m}\u001b[39;49;00m\u001b[38;5;124;43m/A_gene_counts.xlsx\u001b[39;49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[43m)\u001b[49m\n\u001b[1;32m      5\u001b[0m batchA_df\u001b[38;5;241m.\u001b[39mrename(columns\u001b[38;5;241m=\u001b[39m{\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mgene_name\u001b[39m\u001b[38;5;124m'\u001b[39m:\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mSymbol\u001b[39m\u001b[38;5;124m'\u001b[39m},inplace\u001b[38;5;241m=\u001b[39m\u001b[38;5;28;01mTrue\u001b[39;00m)\n\u001b[1;32m      7\u001b[0m \u001b[38;5;66;03m## Create df with relevant locations of gene and promoter + on what chromosome and strand\u001b[39;00m\n",
      "File \u001b[0;32m~/.pyenv/versions/3.12.1/lib/python3.12/site-packages/pandas/io/excel/_base.py:495\u001b[0m, in \u001b[0;36mread_excel\u001b[0;34m(io, sheet_name, header, names, index_col, usecols, dtype, engine, converters, true_values, false_values, skiprows, nrows, na_values, keep_default_na, na_filter, verbose, parse_dates, date_parser, date_format, thousands, decimal, comment, skipfooter, storage_options, dtype_backend, engine_kwargs)\u001b[0m\n\u001b[1;32m    493\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;129;01mnot\u001b[39;00m \u001b[38;5;28misinstance\u001b[39m(io, ExcelFile):\n\u001b[1;32m    494\u001b[0m     should_close \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;01mTrue\u001b[39;00m\n\u001b[0;32m--> 495\u001b[0m     io \u001b[38;5;241m=\u001b[39m \u001b[43mExcelFile\u001b[49m\u001b[43m(\u001b[49m\n\u001b[1;32m    496\u001b[0m \u001b[43m        \u001b[49m\u001b[43mio\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    497\u001b[0m \u001b[43m        \u001b[49m\u001b[43mstorage_options\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mstorage_options\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    498\u001b[0m \u001b[43m        \u001b[49m\u001b[43mengine\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mengine\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    499\u001b[0m \u001b[43m        \u001b[49m\u001b[43mengine_kwargs\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mengine_kwargs\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    500\u001b[0m \u001b[43m    \u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    501\u001b[0m \u001b[38;5;28;01melif\u001b[39;00m engine \u001b[38;5;129;01mand\u001b[39;00m engine \u001b[38;5;241m!=\u001b[39m io\u001b[38;5;241m.\u001b[39mengine:\n\u001b[1;32m    502\u001b[0m     \u001b[38;5;28;01mraise\u001b[39;00m \u001b[38;5;167;01mValueError\u001b[39;00m(\n\u001b[1;32m    503\u001b[0m         \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mEngine should not be specified when passing \u001b[39m\u001b[38;5;124m\"\u001b[39m\n\u001b[1;32m    504\u001b[0m         \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124man ExcelFile - ExcelFile already has the engine set\u001b[39m\u001b[38;5;124m\"\u001b[39m\n\u001b[1;32m    505\u001b[0m     )\n",
      "File \u001b[0;32m~/.pyenv/versions/3.12.1/lib/python3.12/site-packages/pandas/io/excel/_base.py:1550\u001b[0m, in \u001b[0;36mExcelFile.__init__\u001b[0;34m(self, path_or_buffer, engine, storage_options, engine_kwargs)\u001b[0m\n\u001b[1;32m   1548\u001b[0m     ext \u001b[38;5;241m=\u001b[39m \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mxls\u001b[39m\u001b[38;5;124m\"\u001b[39m\n\u001b[1;32m   1549\u001b[0m \u001b[38;5;28;01melse\u001b[39;00m:\n\u001b[0;32m-> 1550\u001b[0m     ext \u001b[38;5;241m=\u001b[39m \u001b[43minspect_excel_format\u001b[49m\u001b[43m(\u001b[49m\n\u001b[1;32m   1551\u001b[0m \u001b[43m        \u001b[49m\u001b[43mcontent_or_path\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mpath_or_buffer\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mstorage_options\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mstorage_options\u001b[49m\n\u001b[1;32m   1552\u001b[0m \u001b[43m    \u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m   1553\u001b[0m     \u001b[38;5;28;01mif\u001b[39;00m ext \u001b[38;5;129;01mis\u001b[39;00m \u001b[38;5;28;01mNone\u001b[39;00m:\n\u001b[1;32m   1554\u001b[0m         \u001b[38;5;28;01mraise\u001b[39;00m \u001b[38;5;167;01mValueError\u001b[39;00m(\n\u001b[1;32m   1555\u001b[0m             \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mExcel file format cannot be determined, you must specify \u001b[39m\u001b[38;5;124m\"\u001b[39m\n\u001b[1;32m   1556\u001b[0m             \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124man engine manually.\u001b[39m\u001b[38;5;124m\"\u001b[39m\n\u001b[1;32m   1557\u001b[0m         )\n",
      "File \u001b[0;32m~/.pyenv/versions/3.12.1/lib/python3.12/site-packages/pandas/io/excel/_base.py:1402\u001b[0m, in \u001b[0;36minspect_excel_format\u001b[0;34m(content_or_path, storage_options)\u001b[0m\n\u001b[1;32m   1399\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;28misinstance\u001b[39m(content_or_path, \u001b[38;5;28mbytes\u001b[39m):\n\u001b[1;32m   1400\u001b[0m     content_or_path \u001b[38;5;241m=\u001b[39m BytesIO(content_or_path)\n\u001b[0;32m-> 1402\u001b[0m \u001b[38;5;28;01mwith\u001b[39;00m \u001b[43mget_handle\u001b[49m\u001b[43m(\u001b[49m\n\u001b[1;32m   1403\u001b[0m \u001b[43m    \u001b[49m\u001b[43mcontent_or_path\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43mrb\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mstorage_options\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mstorage_options\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mis_text\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;28;43;01mFalse\u001b[39;49;00m\n\u001b[1;32m   1404\u001b[0m \u001b[43m\u001b[49m\u001b[43m)\u001b[49m \u001b[38;5;28;01mas\u001b[39;00m handle:\n\u001b[1;32m   1405\u001b[0m     stream \u001b[38;5;241m=\u001b[39m handle\u001b[38;5;241m.\u001b[39mhandle\n\u001b[1;32m   1406\u001b[0m     stream\u001b[38;5;241m.\u001b[39mseek(\u001b[38;5;241m0\u001b[39m)\n",
      "File \u001b[0;32m~/.pyenv/versions/3.12.1/lib/python3.12/site-packages/pandas/io/common.py:882\u001b[0m, in \u001b[0;36mget_handle\u001b[0;34m(path_or_buf, mode, encoding, compression, memory_map, is_text, errors, storage_options)\u001b[0m\n\u001b[1;32m    873\u001b[0m         handle \u001b[38;5;241m=\u001b[39m \u001b[38;5;28mopen\u001b[39m(\n\u001b[1;32m    874\u001b[0m             handle,\n\u001b[1;32m    875\u001b[0m             ioargs\u001b[38;5;241m.\u001b[39mmode,\n\u001b[0;32m   (...)\u001b[0m\n\u001b[1;32m    878\u001b[0m             newline\u001b[38;5;241m=\u001b[39m\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124m\"\u001b[39m,\n\u001b[1;32m    879\u001b[0m         )\n\u001b[1;32m    880\u001b[0m     \u001b[38;5;28;01melse\u001b[39;00m:\n\u001b[1;32m    881\u001b[0m         \u001b[38;5;66;03m# Binary mode\u001b[39;00m\n\u001b[0;32m--> 882\u001b[0m         handle \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;43mopen\u001b[39;49m\u001b[43m(\u001b[49m\u001b[43mhandle\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mioargs\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mmode\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    883\u001b[0m     handles\u001b[38;5;241m.\u001b[39mappend(handle)\n\u001b[1;32m    885\u001b[0m \u001b[38;5;66;03m# Convert BytesIO or file objects passed with an encoding\u001b[39;00m\n",
      "\u001b[0;31mFileNotFoundError\u001b[0m: [Errno 2] No such file or directory: '1. Batch Counts/A_gene_counts.xlsx'"
     ]
    }
   ],
   "source": [
    "DCTRL0_upsigs_df = helper.read_sigs('Up','DCTRL0')\n",
    "DCTRL0_downsigs_df = helper.read_sigs('Down','DCTRL0')\n",
    "\n",
    "batchA_df = pd.read_excel(f'{dirs.batchcounts}/A_gene_counts.xlsx')\n",
    "batchA_df.rename(columns={'gene_name':'Symbol'},inplace=True)\n",
    "\n",
    "## Create df with relevant locations of gene and promoter + on what chromosome and strand\n",
    "genedata_df = batchA_df[['Symbol','gene_start','gene_end','gene_chr','gene_strand']]\n",
    "genedata_df['Promoter_start'] = genedata_df.gene_start - 1000\n",
    "genedata_df['Promoter_end'] = genedata_df.gene_start + 100\n",
    "\n",
    "# subset genedata_df base on significant genes\n",
    "res_up = DCTRL0_upsigs_df.merge(genedata_df,how='inner',on='Symbol')\n",
    "res_down = DCTRL0_downsigs_df.merge(genedata_df,how='inner',on='Symbol')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Symbol</th>\n",
       "      <th>baseMean</th>\n",
       "      <th>log2FoldChange</th>\n",
       "      <th>lfcSE</th>\n",
       "      <th>stat</th>\n",
       "      <th>pvalue</th>\n",
       "      <th>padj</th>\n",
       "      <th>gene_start</th>\n",
       "      <th>gene_end</th>\n",
       "      <th>gene_chr</th>\n",
       "      <th>gene_strand</th>\n",
       "      <th>Promoter_start</th>\n",
       "      <th>Promoter_end</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>DDIT4</td>\n",
       "      <td>2312.933120</td>\n",
       "      <td>3.626941</td>\n",
       "      <td>0.310929</td>\n",
       "      <td>11.664842</td>\n",
       "      <td>1.927613e-31</td>\n",
       "      <td>3.365806e-27</td>\n",
       "      <td>72273920</td>\n",
       "      <td>72276036</td>\n",
       "      <td>10</td>\n",
       "      <td>+</td>\n",
       "      <td>72272920</td>\n",
       "      <td>72274020</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>PFKFB4</td>\n",
       "      <td>1156.457297</td>\n",
       "      <td>2.210382</td>\n",
       "      <td>0.205462</td>\n",
       "      <td>10.758115</td>\n",
       "      <td>5.426893e-27</td>\n",
       "      <td>2.368975e-23</td>\n",
       "      <td>48517684</td>\n",
       "      <td>48562015</td>\n",
       "      <td>3</td>\n",
       "      <td>-</td>\n",
       "      <td>48516684</td>\n",
       "      <td>48517784</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>BHLHE40</td>\n",
       "      <td>656.289464</td>\n",
       "      <td>2.391624</td>\n",
       "      <td>0.221910</td>\n",
       "      <td>10.777457</td>\n",
       "      <td>4.398797e-27</td>\n",
       "      <td>2.368975e-23</td>\n",
       "      <td>4979116</td>\n",
       "      <td>4985323</td>\n",
       "      <td>3</td>\n",
       "      <td>+</td>\n",
       "      <td>4978116</td>\n",
       "      <td>4979216</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>PPFIA4</td>\n",
       "      <td>945.833115</td>\n",
       "      <td>2.614851</td>\n",
       "      <td>0.250851</td>\n",
       "      <td>10.423915</td>\n",
       "      <td>1.928480e-25</td>\n",
       "      <td>6.734637e-22</td>\n",
       "      <td>203026498</td>\n",
       "      <td>203078740</td>\n",
       "      <td>1</td>\n",
       "      <td>+</td>\n",
       "      <td>203025498</td>\n",
       "      <td>203026598</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>MIR210HG</td>\n",
       "      <td>190.890887</td>\n",
       "      <td>3.874204</td>\n",
       "      <td>0.384548</td>\n",
       "      <td>10.074706</td>\n",
       "      <td>7.147440e-24</td>\n",
       "      <td>2.080024e-20</td>\n",
       "      <td>565660</td>\n",
       "      <td>568457</td>\n",
       "      <td>11</td>\n",
       "      <td>-</td>\n",
       "      <td>564660</td>\n",
       "      <td>565760</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1084</th>\n",
       "      <td>AC011472.1</td>\n",
       "      <td>35.521404</td>\n",
       "      <td>1.104751</td>\n",
       "      <td>0.440579</td>\n",
       "      <td>2.507499</td>\n",
       "      <td>1.215890e-02</td>\n",
       "      <td>4.937363e-02</td>\n",
       "      <td>11203628</td>\n",
       "      <td>11216168</td>\n",
       "      <td>19</td>\n",
       "      <td>+</td>\n",
       "      <td>11202628</td>\n",
       "      <td>11203728</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1085</th>\n",
       "      <td>CALCRL</td>\n",
       "      <td>30.278869</td>\n",
       "      <td>1.271135</td>\n",
       "      <td>0.507356</td>\n",
       "      <td>2.505410</td>\n",
       "      <td>1.223097e-02</td>\n",
       "      <td>4.951656e-02</td>\n",
       "      <td>187343129</td>\n",
       "      <td>187448460</td>\n",
       "      <td>2</td>\n",
       "      <td>-</td>\n",
       "      <td>187342129</td>\n",
       "      <td>187343229</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1086</th>\n",
       "      <td>SRP54-AS1</td>\n",
       "      <td>26.816752</td>\n",
       "      <td>1.135093</td>\n",
       "      <td>0.453043</td>\n",
       "      <td>2.505489</td>\n",
       "      <td>1.222823e-02</td>\n",
       "      <td>4.951656e-02</td>\n",
       "      <td>34920858</td>\n",
       "      <td>34982532</td>\n",
       "      <td>14</td>\n",
       "      <td>-</td>\n",
       "      <td>34919858</td>\n",
       "      <td>34920958</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1087</th>\n",
       "      <td>TSHZ2</td>\n",
       "      <td>42.171128</td>\n",
       "      <td>1.166966</td>\n",
       "      <td>0.465843</td>\n",
       "      <td>2.505065</td>\n",
       "      <td>1.224289e-02</td>\n",
       "      <td>4.951889e-02</td>\n",
       "      <td>52972407</td>\n",
       "      <td>53495330</td>\n",
       "      <td>20</td>\n",
       "      <td>+</td>\n",
       "      <td>52971407</td>\n",
       "      <td>52972507</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1088</th>\n",
       "      <td>AP001178.1</td>\n",
       "      <td>13.919111</td>\n",
       "      <td>1.951654</td>\n",
       "      <td>0.779988</td>\n",
       "      <td>2.502158</td>\n",
       "      <td>1.234387e-02</td>\n",
       "      <td>4.981197e-02</td>\n",
       "      <td>653985</td>\n",
       "      <td>657259</td>\n",
       "      <td>18</td>\n",
       "      <td>-</td>\n",
       "      <td>652985</td>\n",
       "      <td>654085</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1089 rows × 13 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "          Symbol     baseMean  log2FoldChange     lfcSE       stat  \\\n",
       "0          DDIT4  2312.933120        3.626941  0.310929  11.664842   \n",
       "1         PFKFB4  1156.457297        2.210382  0.205462  10.758115   \n",
       "2        BHLHE40   656.289464        2.391624  0.221910  10.777457   \n",
       "3         PPFIA4   945.833115        2.614851  0.250851  10.423915   \n",
       "4       MIR210HG   190.890887        3.874204  0.384548  10.074706   \n",
       "...          ...          ...             ...       ...        ...   \n",
       "1084  AC011472.1    35.521404        1.104751  0.440579   2.507499   \n",
       "1085      CALCRL    30.278869        1.271135  0.507356   2.505410   \n",
       "1086   SRP54-AS1    26.816752        1.135093  0.453043   2.505489   \n",
       "1087       TSHZ2    42.171128        1.166966  0.465843   2.505065   \n",
       "1088  AP001178.1    13.919111        1.951654  0.779988   2.502158   \n",
       "\n",
       "            pvalue          padj  gene_start   gene_end gene_chr gene_strand  \\\n",
       "0     1.927613e-31  3.365806e-27    72273920   72276036       10           +   \n",
       "1     5.426893e-27  2.368975e-23    48517684   48562015        3           -   \n",
       "2     4.398797e-27  2.368975e-23     4979116    4985323        3           +   \n",
       "3     1.928480e-25  6.734637e-22   203026498  203078740        1           +   \n",
       "4     7.147440e-24  2.080024e-20      565660     568457       11           -   \n",
       "...            ...           ...         ...        ...      ...         ...   \n",
       "1084  1.215890e-02  4.937363e-02    11203628   11216168       19           +   \n",
       "1085  1.223097e-02  4.951656e-02   187343129  187448460        2           -   \n",
       "1086  1.222823e-02  4.951656e-02    34920858   34982532       14           -   \n",
       "1087  1.224289e-02  4.951889e-02    52972407   53495330       20           +   \n",
       "1088  1.234387e-02  4.981197e-02      653985     657259       18           -   \n",
       "\n",
       "      Promoter_start  Promoter_end  \n",
       "0           72272920      72274020  \n",
       "1           48516684      48517784  \n",
       "2            4978116       4979216  \n",
       "3          203025498     203026598  \n",
       "4             564660        565760  \n",
       "...              ...           ...  \n",
       "1084        11202628      11203728  \n",
       "1085       187342129     187343229  \n",
       "1086        34919858      34920958  \n",
       "1087        52971407      52972507  \n",
       "1088          652985        654085  \n",
       "\n",
       "[1089 rows x 13 columns]"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "res"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Fetch Promoter Sequences"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [],
   "source": [
    "from Bio import Entrez, SeqIO\n",
    "## will take 80 min to go through entire upregulated df\n",
    "#create NCBI Chromosome accession lookup table\n",
    "chromosome_accession_dict = {\n",
    "    '1': 'NC_000001.11',\n",
    "    '2': 'NC_000002.12',\n",
    "    '3': 'NC_000003.12',\n",
    "    '4': 'NC_000004.12',\n",
    "    '5': 'NC_000005.10',\n",
    "    '6': 'NC_000006.12',\n",
    "    '7': 'NC_000007.14',\n",
    "    '8': 'NC_000008.11',\n",
    "    '9': 'NC_000009.12',\n",
    "    '10': 'NC_000010.11',\n",
    "    '11': 'NC_000011.10',\n",
    "    '12': 'NC_000012.12',\n",
    "    '13': 'NC_000013.11',\n",
    "    '14': 'NC_000014.9',\n",
    "    '15': 'NC_000015.10',\n",
    "    '16': 'NC_000016.10',\n",
    "    '17': 'NC_000017.11',\n",
    "    '18': 'NC_000018.10',\n",
    "    '19': 'NC_000019.10',\n",
    "    '20': 'NC_000020.11',\n",
    "    '21': 'NC_000021.9',\n",
    "    '22': 'NC_000022.11',\n",
    "    'X': 'NC_000023.11',\n",
    "    'Y': 'NC_000024.10',\n",
    "}\n",
    "\n",
    "# This function calls entrez fetch function from biopython to fetch sequence that matches region of interest specified\n",
    "def extract_dnasequence(gene_chr,gene_strand,Promoter_start,Promoter_end):\n",
    "\n",
    "    # gene_chr = 3; Promoter_start = 4978116; Promoter_end = 4979216 ; gene_strand = '+'\n",
    "    genestrand_conversion = lambda x: '1' if x =='+' else '2' if x =='+' else 0\n",
    "    # Always provide your email when accessing NCBI services\n",
    "    Entrez.email = \"ae15@stanford.edu\"\n",
    "\n",
    "    # Define the parameters for the sequence you want to fetch\n",
    "    db = \"nucleotide\"  # Database\n",
    "    # Use Entrez.efetch to get the sequence\n",
    "    handle = Entrez.efetch(db='nucleotide', \n",
    "                           id=chromosome_accession_dict[str(gene_chr)], \n",
    "                           rettype=\"fasta\", \n",
    "                           strand= genestrand_conversion(gene_strand),\n",
    "                           seq_start=Promoter_start, \n",
    "                           seq_stop=Promoter_end)\n",
    "\n",
    "    # Parse the fetched sequence\n",
    "    sequence = SeqIO.read(handle, \"fasta\")\n",
    "    handle.close()\n",
    "\n",
    "    # Print the fetched sequence\n",
    "    # print(sequence.id)\n",
    "    # print(sequence.seq)\n",
    "    return sequence.seq\n",
    "\n",
    "extraction_cols = ['gene_chr','gene_strand','Promoter_start','Promoter_end']\n",
    "promoters = []\n",
    "for i in res.index:\n",
    "    input = res[extraction_cols].loc[i].to_list()\n",
    "    if str(input[0]) in chromosome_accession_dict.keys():\n",
    "        promoters.append(extract_dnasequence(*input))\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[18, '-', 653785, 654035]"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "res[extraction_cols].loc[i].to_list()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "# save promoters\n",
    "gene_symbols = res.Symbol.to_list()\n",
    "df = dict(zip(gene_symbols,[str(p) for p in promoters]))\n",
    "promoter_df = pd.DataFrame.from_dict(data=df,orient='index')\n",
    "promoter_df.to_excel(f'{dirs.tfs}/promoters.xlsx')\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Export as FASTA"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "1085"
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "from Bio.Seq import Seq\n",
    "from Bio.SeqRecord import SeqRecord\n",
    "from Bio import SeqIO\n",
    "# Define your sequences\n",
    "\n",
    "gene_symbols = res.Symbol.to_list()\n",
    "\n",
    "promoters_df = pd.read_excel(f'{dirs.tfs}/promoters.xlsx')\n",
    "gene_symbols  = promoters_df[promoters_df.columns[0]].to_list()\n",
    "promoters = promoters_df[promoters_df.columns[1]].to_list()\n",
    "\n",
    "fasta_list = []\n",
    "for i in range(len(promoters)):\n",
    "    fasta_list.append(SeqRecord(Seq(promoters[i]), id=f\"seq{i}\", description=f\"{gene_symbols[i]} promoter region (TSS-1000 to TSS+100)\"))\n",
    "\n",
    "\n",
    "SeqIO.write(fasta_list, f'{dirs.tfs}/promoters.fasta', \"fasta\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## HOMER"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Save significant gene symbols as text"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Symbol</th>\n",
       "      <th>baseMean</th>\n",
       "      <th>log2FoldChange</th>\n",
       "      <th>lfcSE</th>\n",
       "      <th>stat</th>\n",
       "      <th>pvalue</th>\n",
       "      <th>padj</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>PCNX2</td>\n",
       "      <td>937.604003</td>\n",
       "      <td>3.518934</td>\n",
       "      <td>0.103609</td>\n",
       "      <td>33.963677</td>\n",
       "      <td>7.663003e-253</td>\n",
       "      <td>2.451012e-248</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>MFGE8</td>\n",
       "      <td>1089.284818</td>\n",
       "      <td>6.472362</td>\n",
       "      <td>0.199294</td>\n",
       "      <td>32.476525</td>\n",
       "      <td>2.287704e-231</td>\n",
       "      <td>3.658611e-227</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>THY1</td>\n",
       "      <td>2482.720685</td>\n",
       "      <td>10.290578</td>\n",
       "      <td>0.320214</td>\n",
       "      <td>32.136565</td>\n",
       "      <td>1.360802e-226</td>\n",
       "      <td>1.450842e-222</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>DPYSL2</td>\n",
       "      <td>3204.528640</td>\n",
       "      <td>3.917715</td>\n",
       "      <td>0.136434</td>\n",
       "      <td>28.715141</td>\n",
       "      <td>2.468981e-181</td>\n",
       "      <td>1.128148e-177</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>FGFR1</td>\n",
       "      <td>3029.223719</td>\n",
       "      <td>3.808899</td>\n",
       "      <td>0.132898</td>\n",
       "      <td>28.660360</td>\n",
       "      <td>1.190872e-180</td>\n",
       "      <td>4.761257e-177</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8165</th>\n",
       "      <td>SPTBN5</td>\n",
       "      <td>40.584868</td>\n",
       "      <td>0.000891</td>\n",
       "      <td>0.223130</td>\n",
       "      <td>0.003991</td>\n",
       "      <td>9.968156e-01</td>\n",
       "      <td>1.000000e+00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8166</th>\n",
       "      <td>GP6</td>\n",
       "      <td>73.259858</td>\n",
       "      <td>0.000957</td>\n",
       "      <td>0.450079</td>\n",
       "      <td>0.002126</td>\n",
       "      <td>9.983035e-01</td>\n",
       "      <td>1.000000e+00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8167</th>\n",
       "      <td>NPM1P9</td>\n",
       "      <td>12.613348</td>\n",
       "      <td>0.001094</td>\n",
       "      <td>0.428777</td>\n",
       "      <td>0.002551</td>\n",
       "      <td>9.979644e-01</td>\n",
       "      <td>1.000000e+00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8168</th>\n",
       "      <td>EBF2</td>\n",
       "      <td>12.278734</td>\n",
       "      <td>0.001097</td>\n",
       "      <td>0.613201</td>\n",
       "      <td>0.001789</td>\n",
       "      <td>9.985727e-01</td>\n",
       "      <td>1.000000e+00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8169</th>\n",
       "      <td>UCHL3</td>\n",
       "      <td>105.361640</td>\n",
       "      <td>0.000353</td>\n",
       "      <td>0.226791</td>\n",
       "      <td>0.001558</td>\n",
       "      <td>9.987573e-01</td>\n",
       "      <td>1.000000e+00</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>8170 rows × 7 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      Symbol     baseMean  log2FoldChange     lfcSE       stat         pvalue  \\\n",
       "0      PCNX2   937.604003        3.518934  0.103609  33.963677  7.663003e-253   \n",
       "1      MFGE8  1089.284818        6.472362  0.199294  32.476525  2.287704e-231   \n",
       "2       THY1  2482.720685       10.290578  0.320214  32.136565  1.360802e-226   \n",
       "3     DPYSL2  3204.528640        3.917715  0.136434  28.715141  2.468981e-181   \n",
       "4      FGFR1  3029.223719        3.808899  0.132898  28.660360  1.190872e-180   \n",
       "...      ...          ...             ...       ...        ...            ...   \n",
       "8165  SPTBN5    40.584868        0.000891  0.223130   0.003991   9.968156e-01   \n",
       "8166     GP6    73.259858        0.000957  0.450079   0.002126   9.983035e-01   \n",
       "8167  NPM1P9    12.613348        0.001094  0.428777   0.002551   9.979644e-01   \n",
       "8168    EBF2    12.278734        0.001097  0.613201   0.001789   9.985727e-01   \n",
       "8169   UCHL3   105.361640        0.000353  0.226791   0.001558   9.987573e-01   \n",
       "\n",
       "               padj  \n",
       "0     2.451012e-248  \n",
       "1     3.658611e-227  \n",
       "2     1.450842e-222  \n",
       "3     1.128148e-177  \n",
       "4     4.761257e-177  \n",
       "...             ...  \n",
       "8165   1.000000e+00  \n",
       "8166   1.000000e+00  \n",
       "8167   1.000000e+00  \n",
       "8168   1.000000e+00  \n",
       "8169   1.000000e+00  \n",
       "\n",
       "[8170 rows x 7 columns]"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "res = helper.read_results('Up','DCTRL0')\n",
    "res.reset_index(inplace=True)\n",
    "res"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [],
   "source": [
    "symbols_ls = res.Symbol.to_list()\n",
    "with open('my_list.txt', 'w') as file:\n",
    "    for item in symbols_ls:\n",
    "        file.write(str(item) + '\\n')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 204,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "1089"
      ]
     },
     "execution_count": 204,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(symbols_ls)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [
    {
     "ename": "ValueError",
     "evalue": "No records found in handle",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mValueError\u001b[0m                                Traceback (most recent call last)",
      "Cell \u001b[0;32mIn[26], line 20\u001b[0m\n\u001b[1;32m     18\u001b[0m \u001b[38;5;66;03m# Example usage:\u001b[39;00m\n\u001b[1;32m     19\u001b[0m gene_symbol \u001b[38;5;241m=\u001b[39m \u001b[38;5;124m'\u001b[39m\u001b[38;5;124mBRCA1\u001b[39m\u001b[38;5;124m'\u001b[39m  \u001b[38;5;66;03m# Example gene symbol\u001b[39;00m\n\u001b[0;32m---> 20\u001b[0m gene_record \u001b[38;5;241m=\u001b[39m \u001b[43mget_genomic_info\u001b[49m\u001b[43m(\u001b[49m\u001b[43mgene_symbol\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m     21\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m gene_record:\n\u001b[1;32m     22\u001b[0m     \u001b[38;5;28mprint\u001b[39m(gene_record)\n",
      "Cell \u001b[0;32mIn[26], line 14\u001b[0m, in \u001b[0;36mget_genomic_info\u001b[0;34m(gene_symbol)\u001b[0m\n\u001b[1;32m     12\u001b[0m \u001b[38;5;66;03m# Retrieve genomic information for the gene from the Ensembl database\u001b[39;00m\n\u001b[1;32m     13\u001b[0m handle \u001b[38;5;241m=\u001b[39m Entrez\u001b[38;5;241m.\u001b[39mefetch(db\u001b[38;5;241m=\u001b[39m\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mgene\u001b[39m\u001b[38;5;124m'\u001b[39m, \u001b[38;5;28mid\u001b[39m\u001b[38;5;241m=\u001b[39mgene_symbol, rettype\u001b[38;5;241m=\u001b[39m\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mgb\u001b[39m\u001b[38;5;124m'\u001b[39m, retmode\u001b[38;5;241m=\u001b[39m\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mtext\u001b[39m\u001b[38;5;124m'\u001b[39m)\n\u001b[0;32m---> 14\u001b[0m gene_record \u001b[38;5;241m=\u001b[39m \u001b[43mSeqIO\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mread\u001b[49m\u001b[43m(\u001b[49m\u001b[43mhandle\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[38;5;124;43mgenbank\u001b[39;49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[43m)\u001b[49m\n\u001b[1;32m     16\u001b[0m \u001b[38;5;28;01mreturn\u001b[39;00m gene_record\n",
      "File \u001b[0;32m~/.pyenv/versions/3.12.1/lib/python3.12/site-packages/Bio/SeqIO/__init__.py:664\u001b[0m, in \u001b[0;36mread\u001b[0;34m(handle, format, alphabet)\u001b[0m\n\u001b[1;32m    662\u001b[0m     record \u001b[38;5;241m=\u001b[39m \u001b[38;5;28mnext\u001b[39m(iterator)\n\u001b[1;32m    663\u001b[0m \u001b[38;5;28;01mexcept\u001b[39;00m \u001b[38;5;167;01mStopIteration\u001b[39;00m:\n\u001b[0;32m--> 664\u001b[0m     \u001b[38;5;28;01mraise\u001b[39;00m \u001b[38;5;167;01mValueError\u001b[39;00m(\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mNo records found in handle\u001b[39m\u001b[38;5;124m\"\u001b[39m) \u001b[38;5;28;01mfrom\u001b[39;00m \u001b[38;5;28;01mNone\u001b[39;00m\n\u001b[1;32m    665\u001b[0m \u001b[38;5;28;01mtry\u001b[39;00m:\n\u001b[1;32m    666\u001b[0m     \u001b[38;5;28mnext\u001b[39m(iterator)\n",
      "\u001b[0;31mValueError\u001b[0m: No records found in handle"
     ]
    }
   ],
   "source": [
    "from Bio import Entrez\n",
    "from Bio import SeqIO\n",
    "\n",
    "def get_genomic_info(gene_symbol):\n",
    "    Entrez.email = 'ae115@stanford.edu'  # Specify your email for Entrez\n",
    "\n",
    "    # Search for the gene symbol in the Entrez Gene database\n",
    "    handle = Entrez.esearch(db='gene', term=f'{gene_symbol}')\n",
    "    record = Entrez.read(handle)\n",
    "    gene_id = record['IdList'][0]\n",
    "\n",
    "    # Retrieve genomic information for the gene from the Ensembl database\n",
    "    handle = Entrez.efetch(db='gene', id=gene_id, rettype='gb', retmode='text')\n",
    "    gene_record = SeqIO.read(handle, 'genbank')\n",
    "\n",
    "    return gene_record\n",
    "\n",
    "# Example usage:\n",
    "gene_symbol = 'BRCA1'  # Example gene symbol\n",
    "gene_record = get_genomic_info(gene_symbol)\n",
    "if gene_record:\n",
    "    print(gene_record)\n",
    "else:\n",
    "    print(\"Failed to retrieve genomic information for the gene.\")\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.1"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
